File: engine/game_state.py
"""
Game state and effect processor.

GameState holds all runtime state (investigator, current scene/map, message log,
story flags) and processes Effects from the scene/dialogue system.
"""
from __future__ import annotations
import json
import os
from typing import Any, Callable, Dict, List, Optional, TYPE_CHECKING

from .coc import Investigator, check_skill, sanity_check, _parse_dice_notation, Combatant
from .scene import FX, Effect, Scene, SceneRegistry
from .dialogue import DialogueRegistry, DialogueTree, DialogueNode

if TYPE_CHECKING:
    from .tilemap import TileMap


# ---------------------------------------------------------------------------
# Combat encounter definition (lightweight, for scene scripting)
# ---------------------------------------------------------------------------

class CombatEncounter:
    """Simple list of enemies for a combat encounter."""
    def __init__(self, enemies: List[Dict[str, Any]]) -> None:
        self.enemies = enemies  # list of dicts with name, hp, attack_skill, damage_dice...

    def to_combatants(self) -> List[Combatant]:
        result = []
        for e in self.enemies:
            c = Combatant(
                name=e["name"],
                hp=e.get("hp", 10),
                max_hp=e.get("hp", 10),
                san_loss=e.get("san_loss", "0/1d3"),
                dodge_skill=e.get("dodge_skill", 0),
                attack_skill=e.get("attack_skill", 30),
                damage_dice=e.get("damage_dice", "1d6"),
                armor=e.get("armor", 0),
                damage_bonus=e.get("damage_bonus", "0"),
                is_npc=True,
            )
            result.append(c)
        return result


# ---------------------------------------------------------------------------
# Game mode
# ---------------------------------------------------------------------------

class Mode:
    SCENE    = "scene"
    MAP      = "map"
    DIALOGUE = "dialogue"
    COMBAT   = "combat"
    MENU     = "menu"
    GAME_OVER = "game_over"
    VICTORY   = "victory"


# ---------------------------------------------------------------------------
# GameState
# ---------------------------------------------------------------------------

class GameState:
    """
    Central game state.  All game logic flows through this class.
    The renderer reads from it; player input is sent to it.
    """

    def __init__(
        self,
        investigator: Investigator,
        start_scene: str = "start",
    ) -> None:
        self.investigator       = investigator
        self.current_scene_id   = start_scene
        self.mode               = Mode.SCENE
        self.messages: List[str] = []
        self.max_messages       = 200

        # Map state
        self.current_map: Optional["TileMap"] = None
        self.player_x: int = 0
        self.player_y: int = 0

        # Dialogue state
        self.current_dialogue: Optional[DialogueTree] = None
        self.current_node: Optional[DialogueNode] = None

        # Combat state
        self.combat_enemies: List[Combatant] = []
        self.combat_round: int = 0
        self.combat_log: List[str] = []

        # Scene skill check pending push
        self._pending_push_skill: Optional[str] = None

        # Callbacks (set by renderer)
        self.on_combat_start: Optional[Callable] = None
        self.on_victory: Optional[Callable] = None
        self.on_game_over: Optional[Callable] = None

    # -----------------------------------------------------------------------
    # Message log
    # -----------------------------------------------------------------------

    def log(self, msg: str) -> None:
        self.messages.append(msg)
        if len(self.messages) > self.max_messages:
            self.messages = self.messages[-self.max_messages:]

    def recent_messages(self, n: int = 12) -> List[str]:
        return self.messages[-n:]

    # -----------------------------------------------------------------------
    # Scene management
    # -----------------------------------------------------------------------

    def current_scene(self) -> Optional[Scene]:
        return SceneRegistry.get(self.current_scene_id)

    def enter_scene(self, scene_id: str) -> None:
        self.current_scene_id = scene_id
        self.mode = Mode.SCENE
        scene = SceneRegistry.get(scene_id)
        if scene is None:
            self.log(f"[ERROR] Scene not found: {scene_id!r}")
            return
        self.log(f"― {scene.title} ―")
        for effect in scene.on_enter:
            self.apply_effect(effect)

    def process_choice(self, key: str) -> None:
        """Handle player input (a choice key like '1', '2', ...)."""
        scene = self.current_scene()
        if scene is None:
            return
        inv = self.investigator
        choice = scene.find_choice(key, inv)
        if choice is None:
            return

        self.messages = self.messages  # keep log; scene description will redraw
        if choice.skill_check:
            sc = choice.skill_check
            skill_val = inv.skills.get(sc.skill, 0)
            result = inv.roll_skill(sc.skill, sc.bonus_dice)
            self.log(str(result))
            effects = sc.on_success if result.is_success() else sc.on_failure
            for eff in effects:
                self.apply_effect(eff)
        else:
            for eff in choice.effects:
                self.apply_effect(eff)

    # -----------------------------------------------------------------------
    # Dialogue management
    # -----------------------------------------------------------------------

    def start_dialogue(self, dialogue_id: str) -> None:
        tree = DialogueRegistry.get(dialogue_id)
        if tree is None:
            self.log(f"[ERROR] Dialogue not found: {dialogue_id!r}")
            return
        self.current_dialogue = tree
        node = tree.get_node(tree.start_node)
        if node:
            self.current_node = node
            self.mode = Mode.DIALOGUE
            for eff in node.on_enter:
                self.apply_effect(Effect.from_dict(eff))
            self.log(f'{node.speaker}: "{node.text}"')

    def process_dialogue_choice(self, key: str) -> None:
        if not self.current_node or not self.current_dialogue:
            return
        inv = self.investigator
        choices = self.current_node.available_choices(inv)
        try:
            idx = int(key) - 1
            choice = choices[idx]
        except (ValueError, IndexError):
            return

        for eff in choice.effects:
            self.apply_effect(Effect.from_dict(eff))

        if choice.skill_check:
            sc = choice.skill_check
            result = inv.roll_skill(sc.skill)
            self.log(str(result))
            next_id = sc.on_success if result.is_success() else sc.on_failure
        else:
            next_id = choice.next_node

        if next_id:
            node = self.current_dialogue.get_node(next_id)
            if node:
                self.current_node = node
                for eff in node.on_enter:
                    self.apply_effect(Effect.from_dict(eff))
                self.log(f'{node.speaker}: "{node.text}"')
            else:
                self._end_dialogue()
        else:
            self._end_dialogue()

    def _end_dialogue(self) -> None:
        self.current_dialogue = None
        self.current_node = None
        self.mode = Mode.SCENE

    # -----------------------------------------------------------------------
    # Map movement
    # -----------------------------------------------------------------------

    def move_player(self, dx: int, dy: int) -> None:
        if self.current_map is None:
            return
        nx, ny = self.player_x + dx, self.player_y + dy
        if self.current_map.is_passable(nx, ny):
            self.player_x, self.player_y = nx, ny
            self.current_map.reveal_radius(nx, ny, radius=5)
            # Check for entities at new position
            for entity in self.current_map.entities_at(nx, ny):
                if entity.entity_type == "trigger":
                    sid = entity.properties.get("scene_id")
                    if sid:
                        self.enter_scene(sid)
                elif entity.entity_type == "npc":
                    did = entity.properties.get("dialogue_id")
                    if did:
                        self.start_dialogue(did)

    # -----------------------------------------------------------------------
    # Effect processor
    # -----------------------------------------------------------------------

    def apply_effect(self, effect: Effect) -> None:
        t, p = effect.type, effect.params
        inv = self.investigator

        if t == FX.GOTO:
            self.enter_scene(p["scene_id"])

        elif t == FX.GIVE_ITEM:
            inv.add_item(p["name"])
            self.log(f"You obtain: {p['name']}.")

        elif t == FX.REMOVE_ITEM:
            if inv.remove_item(p["name"]):
                self.log(f"You lose: {p['name']}.")

        elif t == FX.DAMAGE:
            amount = p.get("amount", 0)
            if p.get("dice"):
                amount += _parse_dice_notation(p["dice"])
            msgs = inv.take_damage(amount)
            for m in msgs:
                self.log(m)
            if inv.current_hp <= 0:
                self.apply_effect(Effect.game_over(
                    f"{inv.name} succumbs to their wounds."
                ))

        elif t == FX.HEAL:
            amount = p.get("amount", 0)
            if p.get("dice"):
                amount += _parse_dice_notation(p["dice"])
            healed = inv.heal(amount)
            self.log(f"You recover {healed} HP. ({inv.current_hp}/{inv.max_hp})")

        elif t == FX.SET_FLAG:
            inv.set_flag(p["key"], p.get("value", True))

        elif t == FX.CLEAR_FLAG:
            inv.flags.pop(p["key"], None)

        elif t == FX.SAN_LOSS:
            result, lost, msgs = inv.roll_sanity(p["fail_dice"], p.get("success_dice", "0"))
            for m in msgs:
                self.log(m)
            if inv.current_san <= 0:
                self.apply_effect(Effect.game_over("Your mind shatters completely."))

        elif t == FX.GIVE_GOLD:
            inv.cash += p["amount"]
            self.log(f"You gain ${p['amount']}.")

        elif t == FX.TAKE_GOLD:
            inv.cash = max(0, inv.cash - p["amount"])
            self.log(f"You spend ${p['amount']}.")

        elif t == FX.SAY:
            self.log(p["message"])

        elif t == FX.DIALOGUE:
            self.start_dialogue(p["dialogue_id"])

        elif t == FX.LOAD_MAP:
            from .tilemap import TileMap
            try:
                path = f"assets/maps/{p['map_id']}.json"
                tm = TileMap.load(path)
                self.current_map = tm
                px, py = tm.player_start()
                self.player_x, self.player_y = px, py
                tm.reveal_radius(px, py, radius=5)
                self.mode = Mode.MAP
                self.log(f"Entered: {tm.name}")
            except FileNotFoundError:
                self.log(f"[ERROR] Map not found: {p['map_id']}")

        elif t == FX.SHOW_ART:
            # Tell the renderer to display this illustration
            self._pending_art = p.get("illustration_id", "")

        elif t == FX.COMBAT:
            enc = CombatEncounter(p.get("enemies", []))
            self.combat_enemies = enc.to_combatants()
            self.combat_round = 0
            self.combat_log = []
            self.mode = Mode.COMBAT
            names = ", ".join(e.name for e in self.combat_enemies)
            self.log(f"*** COMBAT: {names} ***")
            if self.on_combat_start:
                self.on_combat_start()

        elif t == FX.GAME_OVER:
            msg = p.get("message", "You are dead.")
            self.log(f"GAME OVER: {msg}")
            self.mode = Mode.GAME_OVER
            if self.on_game_over:
                self.on_game_over(msg)

        elif t == FX.VICTORY:
            msg = p.get("message", "You have prevailed.")
            self.log(f"VICTORY: {msg}")
            self.mode = Mode.VICTORY
            if self.on_victory:
                self.on_victory(msg)

    # -----------------------------------------------------------------------
    # Serialization
    # -----------------------------------------------------------------------

    def save(self, path: str) -> None:
        data = {
            "investigator": self.investigator.to_dict(),
            "current_scene": self.current_scene_id,
            "messages": self.messages[-50:],
            "mode": self.mode,
        }
        os.makedirs(os.path.dirname(path), exist_ok=True) if os.path.dirname(path) else None
        with open(path, "w") as f:
            json.dump(data, f, indent=2)
        self.log("Game saved.")

    @classmethod
    def load(cls, path: str) -> "GameState":
        with open(path) as f:
            data = json.load(f)
        inv = Investigator.from_dict(data["investigator"])
        state = cls(inv, start_scene=data.get("current_scene", "start"))
        state.messages = data.get("messages", [])
        state.mode = data.get("mode", Mode.SCENE)
        return state

File: js/engine/game_state.js
// game_state.js – Central game state + effect processor
'use strict';
var DCE = window.DCE || {};
window.DCE = DCE;

DCE.Mode = { SCENE:'scene', MAP:'map', DIALOGUE:'dialogue', COMBAT:'combat',
             MENU:'menu', GAME_OVER:'game_over', VICTORY:'victory' };

DCE.GameState = class {
  constructor(investigator, startScene='intro') {
    this.investigator     = investigator;
    this.currentSceneId   = startScene;
    this.mode             = DCE.Mode.SCENE;
    this.messages         = [];
    this.maxMessages      = 200;

    // Map
    this.currentMap = null;
    this.playerX    = 0;
    this.playerY    = 0;

    // Dialogue
    this.currentDialogue = null;
    this.currentNode     = null;

    // Combat
    this.combatEnemies = [];
    this.combatRound   = 0;

    // Callbacks set by renderer
    this.onUpdate = null;   // called after every state change

    this._pendingArt = null;
  }

  // ---- Message log --------------------------------------------------------
  log(msg) {
    this.messages.push(msg);
    if (this.messages.length > this.maxMessages)
      this.messages = this.messages.slice(-this.maxMessages);
  }

  recentMessages(n=15) { return this.messages.slice(-n); }

  // ---- Scene management --------------------------------------------------
  currentScene() { return DCE.SceneRegistry[this.currentSceneId]; }

  enterScene(sceneId) {
    this.currentSceneId = sceneId;
    this.mode = DCE.Mode.SCENE;
    const scene = DCE.SceneRegistry[sceneId];
    if (!scene) { this.log(`[ERROR] Scene not found: ${sceneId}`); return; }
    this.log(`― ${scene.title} ―`);
    for (const eff of scene.onEnter) this.applyEffect(eff);
    if (this.onUpdate) this.onUpdate();
  }

  processChoice(index) {
    const scene = this.currentScene();
    if (!scene) return;
    const choices = scene.availableChoices(this.investigator);
    const choice = choices[index];
    if (!choice) return;

    if (choice.skillCheck) {
      const sc = choice.skillCheck;
      const result = this.investigator.rollSkill(sc.skill, sc.bonusDice||0);
      this.log(result.toString());
      const effects = result.isSuccess ? sc.onSuccess : sc.onFailure;
      for (const eff of effects) this.applyEffect(eff);
    } else {
      for (const eff of choice.effects) this.applyEffect(eff);
    }
    if (this.onUpdate) this.onUpdate();
  }

  // ---- Dialogue ----------------------------------------------------------
  startDialogue(dialogueId) {
    const tree = DCE.DialogueRegistry[dialogueId];
    if (!tree) { this.log(`[ERROR] Dialogue not found: ${dialogueId}`); return; }
    this.currentDialogue = tree;
    const node = tree.getNode(tree.startNode);
    if (node) {
      this.currentNode = node;
      this.mode = DCE.Mode.DIALOGUE;
      for (const eff of (node.onEnter||[])) this.applyEffect(DCE.Effect.fromDict(eff));
      this.log(`${node.speaker}: "${node.text}"`);
    }
    if (this.onUpdate) this.onUpdate();
  }

  processDialogueChoice(index) {
    if (!this.currentNode || !this.currentDialogue) return;
    const choices = this.currentNode.availableChoices(this.investigator);
    const choice = choices[index];
    if (!choice) return;

    for (const eff of (choice.effects||[])) this.applyEffect(DCE.Effect.fromDict(eff));

    let nextId;
    if (choice.skillCheck) {
      const result = this.investigator.rollSkill(choice.skillCheck.skill);
      this.log(result.toString());
      nextId = result.isSuccess ? choice.skillCheck.onSuccess : choice.skillCheck.onFailure;
    } else {
      nextId = choice.nextNode;
    }

    if (nextId) {
      const node = this.currentDialogue.getNode(nextId);
      if (node) {
        this.currentNode = node;
        for (const eff of (node.onEnter||[])) this.applyEffect(DCE.Effect.fromDict(eff));
        this.log(`${node.speaker}: "${node.text}"`);
      } else { this._endDialogue(); }
    } else { this._endDialogue(); }
    if (this.onUpdate) this.onUpdate();
  }

  _endDialogue() {
    this.currentDialogue = null;
    this.currentNode = null;
    this.mode = DCE.Mode.SCENE;
  }

  // ---- Map movement -------------------------------------------------------
  movePlayer(dx, dy) {
    if (!this.currentMap) return;
    const nx = this.playerX+dx, ny = this.playerY+dy;
    if (this.currentMap.isPassable(nx, ny)) {
      this.playerX = nx; this.playerY = ny;
      this.currentMap.revealRadius(nx, ny, 5);
      for (const ent of this.currentMap.entitiesAt(nx, ny)) {
        if (ent.entityType === 'trigger') {
          const sid = ent.properties?.scene_id;
          if (sid) this.enterScene(sid);
        } else if (ent.entityType === 'npc') {
          const did = ent.properties?.dialogue_id;
          if (did) this.startDialogue(did);
        }
      }
    }
    if (this.onUpdate) this.onUpdate();
  }

  // ---- Effect processor --------------------------------------------------
  applyEffect(effect) {
    const { type, params } = effect;
    const inv = this.investigator;

    if (type === DCE.FX.GOTO) {
      this.enterScene(params.scene_id);

    } else if (type === DCE.FX.GIVE_ITEM) {
      inv.addItem(params.name);
      this.log(`You obtain: ${params.name}.`);

    } else if (type === DCE.FX.REMOVE_ITEM) {
      if (inv.removeItem(params.name)) this.log(`You lose: ${params.name}.`);

    } else if (type === DCE.FX.DAMAGE) {
      let amt = params.amount||0;
      if (params.dice) amt += DCE.parseDice(params.dice);
      const msgs = inv.takeDamage(amt);
      msgs.forEach(m=>this.log(m));
      if (inv.currentHp <= 0)
        this.applyEffect(DCE.Effect.gameOver(`${inv.name} succumbs to their wounds.`));

    } else if (type === DCE.FX.HEAL) {
      const healed = inv.heal(params.amount||0);
      this.log(`Recovered ${healed} HP. (${inv.currentHp}/${inv.maxHp})`);

    } else if (type === DCE.FX.SET_FLAG) {
      inv.setFlag(params.key, params.value !== undefined ? params.value : true);

    } else if (type === DCE.FX.CLEAR_FLAG) {
      delete inv.flags[params.key];

    } else if (type === DCE.FX.SAN_LOSS) {
      const { msgs } = inv.rollSanity(params.fail_dice||'1', params.success_dice||'0');
      msgs.forEach(m=>this.log(m));
      if (inv.currentSan <= 0)
        this.applyEffect(DCE.Effect.gameOver('Your mind shatters completely.'));

    } else if (type === DCE.FX.GIVE_GOLD) {
      inv.cash += params.amount||0;
      this.log(`You gain $${params.amount}.`);

    } else if (type === DCE.FX.TAKE_GOLD) {
      inv.cash = Math.max(0, inv.cash-(params.amount||0));
      this.log(`You spend $${params.amount}.`);

    } else if (type === DCE.FX.SAY) {
      this.log(params.message);

    } else if (type === DCE.FX.DIALOGUE) {
      this.startDialogue(params.dialogue_id);

    } else if (type === DCE.FX.COMBAT) {
      const enemies = (params.enemies||[]).map(e=>new DCE.Combatant({
        name: e.name, hp: e.hp||10, attackSkill: e.attack_skill||30,
        damageDice: e.damage_dice||'1d6', dodgeSkill: e.dodge_skill||0,
        armor: e.armor||0, sanLoss: e.san_loss||'0/1d3',
      }));
      this.combatEnemies = enemies;
      this.combatRound = 0;
      this.mode = DCE.Mode.COMBAT;
      this.log(`*** COMBAT: ${enemies.map(e=>e.name).join(', ')} ***`);

    } else if (type === DCE.FX.GAME_OVER) {
      this.log(`GAME OVER: ${params.message||'You are dead.'}`);
      this.mode = DCE.Mode.GAME_OVER;

    } else if (type === DCE.FX.VICTORY) {
      this.log(`VICTORY: ${params.message||'You have prevailed.'}`);
      this.mode = DCE.Mode.VICTORY;
    }
  }

  // ---- Combat action ------------------------------------------------------
  combatAttack() {
    if (!this.combatEnemies.length) return;
    const enemy = this.combatEnemies[0];
    const msgs = DCE.resolveCombat(this.investigator, enemy);
    msgs.forEach(m=>this.log(m));
    if (enemy.hp <= 0) {
      this.log(`${enemy.name} is defeated!`);
      this.combatEnemies.shift();
      if (!this.combatEnemies.length) {
        this.log('All enemies defeated!');
        this.mode = DCE.Mode.SCENE;
      }
    }
    if (this.investigator.currentHp <= 0)
      this.applyEffect(DCE.Effect.gameOver(`${this.investigator.name} falls in combat.`));
    if (this.onUpdate) this.onUpdate();
  }

  combatFlee() {
    this.log('You flee from combat!');
    this.combatEnemies = [];
    this.mode = DCE.Mode.SCENE;
    if (this.onUpdate) this.onUpdate();
  }

  // ---- Save / Load --------------------------------------------------------
  save() {
    const data = {
      investigator: this.investigator.toDict(),
      currentScene: this.currentSceneId,
      messages: this.messages.slice(-50),
      mode: this.mode,
      playerX: this.playerX,
      playerY: this.playerY,
    };
    localStorage.setItem('dce_save', JSON.stringify(data));
    this.log('Game saved.');
  }

  static load() {
    const raw = localStorage.getItem('dce_save');
    if (!raw) return null;
    const data = JSON.parse(raw);
    const inv = DCE.Investigator.fromDict(data.investigator);
    const state = new DCE.GameState(inv, data.currentScene||'intro');
    state.messages = data.messages||[];
    state.mode = data.mode||DCE.Mode.SCENE;
    state.playerX = data.playerX||0;
    state.playerY = data.playerY||0;
    return state;
  }

  static hasSave() { return !!localStorage.getItem('dce_save'); }
};
